jina/executors/decorators.py

```diff
@@ -470,11 +466,17 @@ def _single(func):
         @wraps(func)
         def arg_wrapper(*args, **kwargs):
             # by default data is in args[1] (self needs to be taken into account)
-            data = args[slice_on]
-
-            if not isinstance(data, Iterable):
+            if (
+                len(args) <= slice_on
+                or isinstance(args[slice_on], str)
+                or isinstance(args[slice_on], bytes)
+                or not isinstance(args[slice_on], Iterable)
+            ):
+                # like this one can use the function with single kwargs
```

Comment:
The following codes lead to en error.
```suggestion
class DummyCrafterTextSingleSliceError(BaseCrafter):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

    @single(slice_on=3)
    def craft(self, text, *args, **kwargs):
        assert isinstance(text, str)
        return {'text': f'{text}-crafted'}
        
def test_batching_text_one_argument_flow(crafter, mocker):
    NUM_DOCS = 15

    def validate_response(resp):
        assert len(resp.index.docs) == NUM_DOCS
        for i, doc in enumerate(resp.index.docs):
            assert doc.text == f'text-{i}-crafted'

    docs = DocumentSet([Document(text=f'text-{i}') for i in range(NUM_DOCS)])
    mock = mocker.Mock()

    with Flow().add(name='crafter', uses='!DummyCrafterTextSingleSliceError') as f:
        f.index(inputs=docs, on_done=mock)

    mock.assert_called_once()
    validate_callback(mock, validate_response)
```

Reply:
I didn't quite catch the conditions we want to check at this point.

Reply:
The conditions are here to ensure that one can still use the crafters with a single instance and not with a list. So that u can still test ur crafter by doing:

result = crafter.craft('text') and see what gets crafted

Reply:
I thought this was a nice feature to mantain